### PR TITLE
More helper functions on state ns

### PR DESCRIPTION
**Pull request by caioaao**
To avoid having ppl complaining about using a monad :speak_no_evil:

### CODE REVIEW

File: src/state_flow/state.clj [hunk 1]
```diff
@@ -95,6 +95,16 @@
...
 
+(defn bind
+  "Equivalent to `(flow \"\" [a mv] (return (f a)))`"
+  [mv f]
+  (m/bind mv f))
+
+(defn fmap
+  "Equivalent to `(flow \"\" [a mv] (f a))`"
+  [f mv]
+  (m/fmap f mv))
```
Comment (philomates) [inline src/state_flow/state.clj]: I find the names `fmap` and `bind` a little abstract, but I guess that is the point :) 
And it looks like they match the haskell names, so it is in line with https://github.com/nubank/state-flow/pull/42

Comment (philomates) [reply src/state_flow/state.clj]: these wrappers aren't really specific to the state-monad, but I guess this is the best place to put them

File: src/state_flow/state.clj [hunk 2]
```diff
@@ -95,6 +95,16 @@
                  (d/pair (my-fn) s))
                error-context))
 
+(defn bind
+  "Equivalent to `(flow \"\" [a mv] (return (f a)))`"
+  [mv f]
+  (m/bind mv f))
+
+(defn fmap
+  "Equivalent to `(flow \"\" [a mv] (f a))`"
```
Comment (philomates) [inline src/state_flow/state.clj]: I think a prose description would also be helpful (along with tests that serve as examples)

File: src/state_flow/state.clj (hunk 1 above)
Comment (caioaao) [reply src/state_flow/state.clj]: about the names: yeah, I couldn't find better names either, and at least people that used spec/core.check generators will be familiar with them.

about being specific to the state monad: they're not, but that's the point: passing the impression of an specific/unified API for dealing with state-flow's states and more advanced users can drop down to the monad/functors/etc api if they want to.

File: src/state_flow/state.clj (hunk 2 above)
Comment (caioaao) [reply src/state_flow/state.clj]: I could use some help on this tbh :sweat_smile: . The equivalent is not even strictly true, but it was the best way I found to describe it with only state-flow's lingo

Comment (philomates) [reply src/state_flow/state.clj]: maybe adapting what is in cats?
https://github.com/funcool/cats/blob/master/src/cats/core.cljc#L82-L91

File: src/state_flow/state.clj [hunk 3]
```diff
@@ -95,6 +95,16 @@
                  (d/pair (my-fn) s))
                error-context))
 
+(defn bind
+  "Equivalent to `(flow \"\" [a mv] (return (f a)))`"
```
Comment (dchelimsky) [inline src/state_flow/state.clj]: There's no argument `a` below, so I'm not clear what this means.

Comment (dchelimsky) [reply src/state_flow/state.clj]: Oh, never mind - we're binding the result of `mv` to `a`. This could definitely be made more clear for the monadic-impaired.

Comment (caioaao) [reply src/state_flow/state.clj]: Oh yeah. You not getting it on the first read is already convincing that this explanation is bad :)

Comment (dchelimsky) [reply src/state_flow/state.clj]: Can you show an example of how to use it?

Comment (caioaao) [reply src/state_flow/state.clj]: Sure! The biggest usage in my case is this: I have a function that creates an user and returns it:
```clojure
(defn user-signup! "Signs up using parameters and return user" [parameters]
  (flow "Creating complete user"
        [user (http/create! parameters)]
        (confirm-email! user)
        (cats/return user)))
```

Sometimes I need to have the returned user during a flow, like this
```clojure
(defflow bla
  ...
  [user (user-after-signup! {...})]
  ;; use unwrapped user value here
  )
```

But sometimes I just want to pass it to another function. Without the `bind` I'd have to do this:
```clojure
(defflow ble
  [user (user-after-signup! {...})
   foo (use-user! user)]
   ...
   )
```

But with bind:
```clojure
(defflow ble-2
  [foo (-> (user-after-signup! {...})
           (bind use-user!))]
   ...
  )
```

So basically a helper to compose functions that evaluate to monadic values (there's even an arrow macro equivalent on cats - `cats/->=` - but that's too much cognitive load for a testing library in my opinion)

Comment (dchelimsky) [reply src/state_flow/state.clj]: Thanks, that helps.

Comment (dchelimsky) [reply src/state_flow/state.clj]: OK, so the description is somewhere between "here's what this does/returns" and "here's how you use it", but not quite either. I'll follow up with suggestions.

Side note: the `user-signup!` function, I really want to discourage using `cats` directly. There is a `state-flow.state/return` function that calls into `cats.core/return` and I think that's what we should be using. WDYT?

Comment (dchelimsky) [reply src/state_flow/state.clj]: I think `bind` is a confusing name here because we're really applying a function to an unwrapped value, which we _then_ bind to the symbol on the left. You say elsewhere that `cats/->=` is too much cognitive overload, but I think this would be easier to grok than your example above:


```clojure
(defflow ble-2
  [foo (->= (user-after-signup! {...})
             use-user!))]
   ...
  )
```

Alternatively, we could name it something that relates more to state-flow, but still has the arrow in it, like `state/->`.

Comment (caioaao) [reply src/state_flow/state.clj]: Oh, now that you said it, I like the `state/->` and it'd make `bind` redundant. Looks a lot clearer and even more useful than just `bind` itself :+1:

**Comment by sovelten** [comment]
I'm not really sure this should live on the `state` namespace or some other namespace.
For instance, you mention flows in the docs, but `state.clj` namespace is not aware of flow, it is the opposite: flow requires state.

How about moving to some other namespace like helpers.clj or something like that?

**Comment by caioaao** [comment]
> I'm not really sure this should live on the `state` namespace or some other namespace.
> For instance, you mention flows in the docs, but `state.clj` namespace is not aware of flow, it is the opposite: flow requires state.
> 
> How about moving to some other namespace like helpers.clj or something like that?

I think that's more a reflection on the bad docstrings than actually being at the wrong place. I used flows here because I thought it was the more familiar concept I could use. We can revisit the placement after docstrings are better, wdyt?

File: src/state_flow/state.clj [hunk 4]
```diff
@@ -59,27 +60,27 @@
 (util/make-printable (type error-context))
 
 (defn get
-  "Returns the equivalent of (fn [state] [state, state])"
+  "Evaluates the equivalent of (fn [state] [state, state])"
```
Comment (dchelimsky) [inline src/state_flow/state.clj]: Returns is correct. Why did you change this?

Comment (dchelimsky) [reply src/state_flow/state.clj]: What it returns is a state monad, which wraps a function like `(fn [state] [state, state])`, but there is no evaluation when you invoke `(get)`.

File: src/state_flow/state.clj [hunk 5]
```diff
@@ -95,6 +96,47 @@
...
                error-context))
 
+(defmacro ->
+  "Analog to Clojure's `->`, but operating on expressions that result in a state
+  instance. An example:
+
+  Usage:
+
+  ```clojure
+  (defn inc [x] (state/return (inc x)))
```
Comment (dchelimsky) [inline src/state_flow/state.clj]: Please call this something other than a core clojure function :) Maybe `inc-state` or some such.

File: src/state_flow/state.clj [hunk 6]
```diff
@@ -95,6 +96,47 @@
                  (d/pair (my-fn) s))
                error-context))
 
+(defmacro ->
+  "Analog to Clojure's `->`, but operating on expressions that result in a state
+  instance. An example:
+
+  Usage:
+
+  ```clojure
```
Comment (dchelimsky) [inline src/state_flow/state.clj]: I haven't seen this markdown syntax used within docstrings. What environment formats this for you? Not github, and not a REPL. Where else?

File: src/state_flow/state.clj [hunk 7]
```diff
@@ -95,6 +96,47 @@
                  (d/pair (my-fn) s))
                error-context))
 
+(defmacro ->
```
Comment (sovelten) [inline src/state_flow/state.clj]: This is equivalent to `(fmap inc (state/return 1))`
I never use this kind of macro and I would not want to encourage anyone to use it. It is better to have one clear way to do it than several different ways.

If you want to apply several functions, `comp` works fine:
`(m/fmap (comp inc inc) (state/return 1))`

And there is no need to differentiate between `->` and `->>` as `comp` is agnostic to that and it receives functions as arguments and not pieces of code with an argument missing. No need to introduce thread macros to the monad world as thread macros are a bad design choice anyway.

Comment (dchelimsky) [reply src/state_flow/state.clj]: > thread macros are a bad design choice anyway

That's a pretty objective sounding statement for a subjective statement ;) Care to elaborate?

Comment (dchelimsky) [reply src/state_flow/state.clj]: I do agree, btw, with excluding `->` and `->>` here. Just want to understand why you think they're universally bad.

Comment (sovelten) [reply src/state_flow/state.clj]: I have just the thing: https://gist.github.com/sovelten/486602d2711ed251ffee41215b7e8dbc

Comment (caioaao) [reply src/state_flow/state.clj]: heh, that was a test. I knew you wouldn't like it @sovelten :joy: But `comp` doesn't work when each functions is already returning a monadic value. we'd need something like haskell's `>=` (which is ugly and confusing af) or re-add `bind` here (which is not the preferred way even in haskell). This, IMO, is a clear place where a new syntax can/should be introduced and what I did was just introducing one that would already be familiar to Clojure users

File: src/state_flow/state.clj (hunk 6 above)
Comment (caioaao) [reply src/state_flow/state.clj]: I think cljdoc.org supports this, but now that you mention I don't know if it's true. Anyway, I'd say this is not the optimal way of adding examples (it's not testable as in Python), but I couldn't think of an alternative that would have it close to the function definition.

File: src/state_flow/state.clj (hunk 4 above)
Comment (caioaao) [reply src/state_flow/state.clj]: Oops. I wasn't supposed to commit that, sorry :sweat: . I was rewording it from something I've read a long time ago about how the lingo should be different for REPL based languages, but I couldn't find it and wasn't sure about the discussion anymore.

File: src/state_flow/state.clj (hunk 6 above)
Comment (dchelimsky) [reply src/state_flow/state.clj]: Just leave off the markdown (i.e. `\``clojure ... ```) and it's fine to have the Usage example here IMO.

File: src/state_flow/state.clj (hunk 7 above)
Comment (sovelten) [reply src/state_flow/state.clj]: Ok, I get the idea now. But I don't see the benefit of yet another syntactic sugar option, it is just another thing for people to learn and get confused about. `do-let` already provides all the necessary flexibility and it mirrors Haskell do notation.

Comment (caioaao) [reply src/state_flow/state.clj]: You're right the library already provides enough flexibility, but it does lead to a pretty verbose code IMO. This might be a matter of preference/style, though.

File: src/state_flow/state.clj (hunk 6 above)
Comment (caioaao) [reply src/state_flow/state.clj]: Just double-checked and cljdoc.org renders the markdown in docstrings. Might be better to leave as is then, no?

File: src/state_flow/state.clj (hunk 7 above)
Comment (dchelimsky) [reply src/state_flow/state.clj]: @caioaao we want to be cautious about expanding the surface of state-flow, so we're not going to add these right now. We think you're not alone in wanting to reduce verbosity, and that an open source extension library would be appreciated by other users of state-flow, so that's definitely an option for you. Alternatively, you could just use the cats macros and functions directly (`flow` is just a wrapper for `do-let`). I'll add something to the README saying "For advanced users,`flow` is a wrapper for `cats/do-let`, so you can use any `cats` functions and macros that you'd use within a `do-let` within a `flow`".
